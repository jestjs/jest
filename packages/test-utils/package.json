{
  "name": "@jest/test-utils",
  "version": "30.0.0-alpha.7",
  "private": true,
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "exports": {
    ".": {
      "types": "./build/index.d.ts",
      "require": "./build/index.js",
      "import": "./build/index.mjs",
      "default": "./build/index.js"
    },
    "./package.json": "./package.json"
  },
  "dependencies": {
    "@jest/globals": "workspace:*",
    "@jest/pattern": "workspace:*",
    "@jest/types": "workspace:*",
    "@types/node": "*",
    "ansi-regex": "^5.0.1",
    "ansi-styles": "^5.0.0",
    "pretty-format": "workspace:*",
    "semver": "^7.5.3"
  },
  "devDependencies": {
    "@types/semver": "^7.1.0"
  },
  "engines": {
    "node": "^16.10.0 || ^18.12.0 || >=20.0.0"
  }
}
