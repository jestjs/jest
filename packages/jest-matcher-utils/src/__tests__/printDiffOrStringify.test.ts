/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {alignedAnsiStyleSerializer} from '@jest/test-utils';
import {INVERTED_COLOR, printDiffOrStringify} from '../index';

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

describe('printDiffOrStringify', () => {
  const testDiffOrStringify = (expected: unknown, received: unknown): string =>
    printDiffOrStringify(expected, received, 'Expected', 'Received', true);

  test('expected is empty and received is single line', () => {
    const expected = '';
    const received = 'single line';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected is multi line and received is empty', () => {
    const expected = 'multi\nline';
    const received = '';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected and received are single line with multiple changes', () => {
    const expected = 'delete common expected common prev';
    const received = 'insert common received common next';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected and received are multi line with trailing spaces', () => {
    const expected = 'delete \ncommon expected common\nprev ';
    const received = 'insert \ncommon received common\nnext ';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('has no common after clean up chaff multiline', () => {
    const expected = 'delete\ntwo';
    const received = 'insert\n2';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('has no common after clean up chaff one-line', () => {
    const expected = 'delete';
    const received = 'insert';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('object contain readonly symbol key object', () => {
    const expected = {b: 2};
    const received = {b: 1};
    const symbolKey = Symbol.for('key');
    Object.defineProperty(expected, symbolKey, {
      configurable: true,
      enumerable: true,
      value: {
        a: 1,
      },
      writable: false,
    });
    Object.defineProperty(received, symbolKey, {
      configurable: true,
      enumerable: true,
      value: {
        a: 1,
      },
      writable: false,
    });
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  describe('MAX_DIFF_STRING_LENGTH', () => {
    const lessChange = INVERTED_COLOR('single ');
    const less = 'single line';
    const more = `multi line${'\n123456789'.repeat(2000)}`; // 10 + 20K chars

    test('both are less', () => {
      const difference = testDiffOrStringify('multi\nline', less);

      expect(difference).toMatch('- multi');
      expect(difference).toMatch('- line');

      // diffStringsUnified has substring change
      expect(difference).not.toMatch('+ single line');
      expect(difference).toMatch(lessChange);
    });

    test('expected is more', () => {
      const difference = testDiffOrStringify(more, less);

      expect(difference).toMatch('- multi line');
      expect(difference).toMatch('+ single line');

      // diffLinesUnified does not have substring change
      expect(difference).not.toMatch(lessChange);
    });

    test('received is more', () => {
      const difference = testDiffOrStringify(less, more);

      expect(difference).toMatch('- single line');
      expect(difference).toMatch('+ multi line');

      // diffLinesUnified does not have substring change
      expect(difference).not.toMatch(lessChange);
    });
  });

  describe('asymmetricMatcher', () => {
    test('minimal test', () => {
      const expected = {a: expect.any(Number), b: 2};
      const received = {a: 1, b: 1};
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('jest asymmetricMatcher', () => {
      const expected = {
        a: expect.any(Number),
        b: expect.anything(),
        c: expect.arrayContaining([1, 3]),
        d: 'jest is awesome',
        e: 'jest is awesome',
        f: {
          a: new Date(),
          b: 'jest is awesome',
        },
        g: true,
        [Symbol.for('h')]: 'jest is awesome',
      };
      const received = {
        a: 1,
        b: 'anything',
        c: [1, 2, 3],
        d: expect.stringContaining('jest'),
        e: expect.stringMatching(/^jest/),
        f: expect.objectContaining({
          a: expect.any(Date),
        }),
        g: false,
        [Symbol.for('h')]: expect.any(String),
      };

      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('custom asymmetricMatcher', () => {
      expect.extend({
        equal5(received: unknown) {
          if (received === 5)
            return {
              message: () => `expected ${received} not to be 5`,
              pass: true,
            };
          return {
            message: () => `expected ${received} to be 5`,
            pass: false,
          };
        },
      });
      const expected = {
        a: expect.equal5(),
        b: false,
      };
      const received = {
        a: 5,
        b: true,
      };

      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('nested object', () => {
      const expected = {
        a: 1,
        b: {
          a: 1,
          b: expect.any(Number),
        },
        c: 2,
      };
      const received = {
        a: expect.any(Number),
        b: {
          a: 1,
          b: 2,
        },
        c: 1,
      };
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('array', () => {
      const expected: Array<unknown> = [1, expect.any(Number), 3];
      const received: Array<unknown> = [1, 2, 2];
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('object in array', () => {
      const expected: Array<unknown> = [1, {a: 1, b: expect.any(Number)}, 3];
      const received: Array<unknown> = [1, {a: 1, b: 2}, 2];
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('map', () => {
      const expected = new Map<string, unknown>([
        ['a', 1],
        ['b', expect.any(Number)],
        ['c', 3],
      ]);
      const received = new Map<string, unknown>([
        ['a', 1],
        ['b', 2],
        ['c', 2],
      ]);
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('circular object', () => {
      const expected: any = {
        b: expect.any(Number),
        c: 3,
      };
      expected.a = expected;
      const received: any = {
        b: 2,
        c: 2,
      };
      received.a = received;
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('transitive circular', () => {
      const expected: unknown = {
        a: 3,
      };
      expected.nested = {b: expect.any(Number), parent: expected};
      const received: any = {
        a: 2,
      };
      received.nested = {b: 2, parent: received};
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('circular array', () => {
      const expected: Array<unknown> = [1, expect.any(Number), 3];
      expected.push(expected);
      const received: Array<unknown> = [1, 2, 2];
      received.push(received);
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });

    test('circular map', () => {
      const expected = new Map<string, unknown>([
        ['a', 1],
        ['b', expect.any(Number)],
        ['c', 3],
      ]);
      expected.set('circular', expected);
      const received = new Map<string, unknown>([
        ['a', 1],
        ['b', 2],
        ['c', 2],
      ]);
      received.set('circular', received);
      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });
  });

  describe('getters', () => {
    test('handles self-referential getters without infinite recursion', () => {
      class TestClass {
        constructor(public value: string) {}

        get selfRef() {
          return new TestClass(`${this.value}_ref`);
        }
      }

      const expected = new TestClass('hello');
      const received = new TestClass('world');

      expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
    });
  });
});
