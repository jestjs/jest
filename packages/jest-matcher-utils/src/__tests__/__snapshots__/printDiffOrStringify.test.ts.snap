// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`printDiffOrStringify asymmetricMatcher array 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Array [</>
<d>    1,</>
<d>    Any<Number>,</>
<g>-   3,</>
<r>+   2,</>
<d>  ]</>
`;

exports[`printDiffOrStringify asymmetricMatcher circular array 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Array [</>
<d>    1,</>
<d>    Any<Number>,</>
<g>-   3,</>
<r>+   2,</>
<d>    [Circular],</>
<d>  ]</>
`;

exports[`printDiffOrStringify asymmetricMatcher circular map 1`] = `
<g>- Expected  - 2</>
<r>+ Received  + 2</>

<d>  Map {</>
<d>    "a" => 1,</>
<d>    "b" => Any<Number>,</>
<g>-   "c" => 3,</>
<r>+   "c" => 2,</>
<d>    "circular" => Map {</>
<d>      "a" => 1,</>
<d>      "b" => Any<Number>,</>
<g>-     "c" => 3,</>
<r>+     "c" => 2,</>
<d>      "circular" => [Circular],</>
<d>    },</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher circular object 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<d>    "a": [Circular],</>
<d>    "b": Any<Number>,</>
<g>-   "c": 3,</>
<r>+   "c": 2,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher custom asymmetricMatcher 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<d>    "a": equal5<>,</>
<g>-   "b": false,</>
<r>+   "b": true,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher jest asymmetricMatcher 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<d>    "a": Any<Number>,</>
<d>    "b": Anything,</>
<d>    "c": ArrayContaining [</>
<d>      1,</>
<d>      3,</>
<d>    ],</>
<d>    "d": StringContaining "jest",</>
<d>    "e": StringMatching /^jest/,</>
<d>    "f": ObjectContaining {</>
<d>      "a": Any<Date>,</>
<d>    },</>
<g>-   "g": true,</>
<r>+   "g": false,</>
<d>    Symbol(h): Any<String>,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher map 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Map {</>
<d>    "a" => 1,</>
<d>    "b" => Any<Number>,</>
<g>-   "c" => 3,</>
<r>+   "c" => 2,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher minimal test 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<d>    "a": Any<Number>,</>
<g>-   "b": 2,</>
<r>+   "b": 1,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher nested object 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<d>    "a": Any<Number>,</>
<d>    "b": Object {</>
<d>      "a": 1,</>
<d>      "b": Any<Number>,</>
<d>    },</>
<g>-   "c": 2,</>
<r>+   "c": 1,</>
<d>  }</>
`;

exports[`printDiffOrStringify asymmetricMatcher object in array 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Array [</>
<d>    1,</>
<d>    Object {</>
<d>      "a": 1,</>
<d>      "b": Any<Number>,</>
<d>    },</>
<g>-   3,</>
<r>+   2,</>
<d>  ]</>
`;

exports[`printDiffOrStringify asymmetricMatcher transitive circular 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<g>-   "a": 3,</>
<r>+   "a": 2,</>
<d>    "nested": Object {</>
<d>      "b": Any<Number>,</>
<d>      "parent": [Circular],</>
<d>    },</>
<d>  }</>
`;

exports[`printDiffOrStringify expected and received are multi line with trailing spaces 1`] = `
<g>- Expected  - 3</>
<r>+ Received  + 3</>

<g>- <i>delete</i><Y> </></>
<r>+ <i>insert</i><Y> </></>
<g>- common <i>expect</i>ed common</>
<r>+ common <i>receiv</i>ed common</>
<g>- <i>prev</i><Y> </></>
<r>+ <i>next</i><Y> </></>
`;

exports[`printDiffOrStringify expected and received are single line with multiple changes 1`] = `
Expected: <g>"<i>delete</i> common <i>expect</i>ed common <i>prev</i>"</>
Received: <r>"<i>insert</i> common <i>receiv</i>ed common <i>next</i>"</>
`;

exports[`printDiffOrStringify expected is empty and received is single line 1`] = `
Expected: <g>""</>
Received: <r>"single line"</>
`;

exports[`printDiffOrStringify expected is multi line and received is empty 1`] = `
Expected: <g>"multi</>
<g>line"</>
Received: <r>""</>
`;

exports[`printDiffOrStringify has no common after clean up chaff multiline 1`] = `
<g>- Expected  - 2</>
<r>+ Received  + 2</>

<g>- delete</>
<g>- two</>
<r>+ insert</>
<r>+ 2</>
`;

exports[`printDiffOrStringify has no common after clean up chaff one-line 1`] = `
Expected: <g>"delete"</>
Received: <r>"insert"</>
`;

exports[`printDiffOrStringify object contain readonly symbol key object 1`] = `
<g>- Expected  - 1</>
<r>+ Received  + 1</>

<d>  Object {</>
<g>-   "b": 2,</>
<r>+   "b": 1,</>
<d>    Symbol(key): Object {</>
<d>      "a": 1,</>
<d>    },</>
<d>  }</>
`;
